src/librawspeed/decompressors/LJpegDecompressor.cpp

```diff
@@ -217,8 +249,37 @@ template <int N_COMP, bool WeirdWidth> void LJpegDecompressor::decodeN() {
       assert(trailingPixels < N_COMP);
       unsigned c = 0;
       for (; c < trailingPixels; ++c) {
-        pred[c] = uint16_t(pred[c] + ht[c]->decodeDifference(bitStream));
-        img(row, col + c) = pred[c];
+        // Continue predictor update skipped at last full block
+        int predA = img(row, col - N_COMP + c);
+        int predB = predMode > 1 ? img(row - 1, col + c) : 0;
+        int predC = predMode > 1 ? img(row - 1, col - N_COMP + c) : 0;
+        switch (predMode) {
+        case 1:
+          pred[c] = uint16_t(predA);
+          break;
+        case 2:
+          pred[c] = uint16_t(predB);
+          break;
+        case 3:
+          pred[c] = uint16_t(predC);
+          break;
+        case 4:
+          pred[c] = uint16_t(predA + predB - predC);
```

Comment:
Predictor 4 is special because A + B - C can overflow uint16_t. The prediction and diff calculation must be done in int32_t. Prediction value can be negative, too.

Reply:
Yep, that's why predA, predB, predC are declared as ints above. Are you saying I should check and clamp to 0 here as well?

Reply:
The reconstruction using diff and prediction value must be done with signed integers, then you can safely cast the result back to uint16_t. If you cast the prediction value first (loosing some bits) and then do simple unsigned integer math, I assume you get wrong results.

Reply:
My reading of itu-t81 is that only the intermediate calculation is to be done as signed int. The predictor is stored and used "modulo 2^16", i.e. unsigned:

> The difference between the prediction value and the input is calculated modulo 2^16. In the decoder the difference is
decoded and added, modulo 2^16, to the prediction.

But I do need to check how the two are added back together earlier indeed.

Reply:
I _think_ all the uint16_t casting takes care of the modulo operations described above by design, and no changes are in fact necessary, but happy to be proven wrong.

Reply:
Well that's difficult to prove. I've added a test case to my implementation where `A = B = 65535` and `C=0` and the compression/decompression verification fails if I calculate this with u16 instead of i32. I can not tell you if some integer magic solves this in your code but it seems legit to assume there is an overflow which leads to wrong results in edge cases.

For me, the edge case was linear raw data (RGB instead of CFA) that uses the full 2^16 value range, writing into DNG with ljpeg compression. For regular CFA images, the input is ~ 2^14, so you never hit overflows.

It's difficult to test as we don't have a ljpeg compressor in rawspeed to write such small tests.